packages/glob/src/glob.ts

```diff
@@ -0,0 +1,189 @@
+        continue
+      }
+
+      // Match
+      const matchResult = patternHelper.match(patterns, item.path)
+
+      // Directory
+      if (stats.isDirectory()) {
+        // Matched
+        if (matchResult & MatchResult.Directory) {
+          result.push(item.path)
+        }
+        // Descend?
+        else if (!patternHelper.partialMatch(patterns, item.path)) {
+          continue
+        }
+
+        // Push the child items in reverse
+        const childLevel = item.level + 1
+        const childItems = (await fs.promises.readdir(item.path)).map(
+          x => new SearchState(path.join(item.path, x), childLevel)
+        )
+        stack.push(...childItems.reverse())
+      }
+      // File
+      else if (matchResult & MatchResult.File) {
+        result.push(item.path)
+      }
+    }
+  }
+
+  return result
+}
+
+/**
+ * Returns the search path preceeding the first segment that contains a pattern.
+ *
+ * For example, '/foo/bar*' returns '/foo'.
+ */
+export function getSearchPath(pattern: string): string {
+  const patterns: Pattern[] = patternHelper.parse([pattern], getOptions())
+  const searchPaths: string[] = patternHelper.getSearchPaths(patterns)
+  return searchPaths.length > 0 ? searchPaths[0] : ''
+}
+
+async function stat(
+  item: SearchState,
+  options: IGlobOptions,
+  traversalChain: string[]
+): Promise<fs.Stats | undefined> {
+  // Note:
+  // `stat` returns info about the target of a symlink (or symlink chain)
+  // `lstat` returns info about a symlink itself
+  let stats: fs.Stats
+  if (options.followSymbolicLinks) {
+    try {
+      // Use `stat` (following symlinks)
+      stats = await fs.promises.stat(item.path)
+    } catch (err) {
+      if (err.code === 'ENOENT') {
+        if (options.omitBrokenSymbolicLinks) {
+          core.debug(`Broken symlink '${item.path}'`)
+          return undefined
+        }
+
+        throw new Error(
+          `No information found for the path '${
+            item.path
+          }'. This may indicate a broken symbolic link.`
+        )
+      }
+
+      throw err
+    }
+  } else {
+    // Use `lstat` (not following symlinks)
+    stats = await fs.promises.lstat(item.path)
+  }
+
+  // Note, isDirectory() returns false for the lstat of a symlink
+  if (stats.isDirectory() && options.followSymbolicLinks) {
+    // Get the realpath
+    const realPath: string = await fs.promises.realpath(item.path)
+
+    // Fixup the traversal chain to match the item level
+    while (traversalChain.length >= item.level) {
+      traversalChain.pop()
+    }
+
+    // Test for a cycle
+    if (traversalChain.some((x: string) => x === realPath)) {
+      core.debug(
+        `Symlink cycle detected for path '${
+          item.path
+        }' and realpath '${realPath}'`
+      )
+      return undefined
+    }
+
+    // Update the traversal chain
+    traversalChain.push(realPath)
+  }
+
+  return stats
+}
+
+function getOptions(copy?: IGlobOptions): IGlobOptions {
```

Comment:
Why not pass the glob options directly to the functions instead of copying?

Reply:
The properties are optional and the defaults need to be set when undefined. As an SDK function, it shouldnt mutate the object passed in by the caller. Hence the copy.